### PR TITLE
parser: implement Restore for FrameBound

### PR DESCRIPTION
**Pull request by exialin**
Implement `Restore` and unit test for `FrameBound`.

I changed two lines in `parser.y`, though not sure it is completely right.

Issue: https://github.com/pingcap/tidb/issues/8532

### CODE REVIEW

File: ast/dml.go [hunk 1]
```diff
@@ -1519,7 +1519,41 @@ type FrameBound struct {
...
+			if err := n.Expr.Restore(ctx); err != nil {
+				return errors.Annotate(err, "An error occurred while restore FrameBound.Expr")
+			}
+		}
+		if n.Unit != nil {
+			// Here the Unit string should not be quoted
+			singleQuotesFlag := ctx.Flags & RestoreStringSingleQuotes
+			doubleQuotesFlag := ctx.Flags & RestoreStringDoubleQuotes
+			ctx.Flags &= ^RestoreStringSingleQuotes
+			ctx.Flags &= ^RestoreStringDoubleQuotes
```
Comment (kennytm) [inline ast/dml.go]: This doesn't seem like the correct approach. For instance, the following function call shouldn't parse either:

```sql
select date_add(current_timestamp, interval 1 'year');
```

which means the same logic got to be duplicated for FuncCallExpr.

----

I have no idea why `Unit` is considered a *string literal* expression?! It should be made a special "unit expression" based on an enumerated constant. That involves changing TiDB itself though.

Comment (exialin) [reply ast/dml.go]: Yes, I think it is reasonable. And without this change, it is harder for `FuncCallExpr` to handle this kind of cases.

File: ast/dml.go [hunk 2]
```diff
@@ -1519,7 +1519,41 @@ type FrameBound struct {
...
+			ctx.WriteKeyWord("INTERVAL ")
+		}
+		if n.Expr != nil {
+			if err := n.Expr.Restore(ctx); err != nil {
+				return errors.Annotate(err, "An error occurred while restore FrameBound.Expr")
+			}
+		}
+		if n.Unit != nil {
+			// Here the Unit string should not be quoted
+			singleQuotesFlag := ctx.Flags & RestoreStringSingleQuotes
```
Comment (leoppro) [inline ast/dml.go]: My temporary plan:
```
var sb string.Builder
n.Unit.Restore(NewRestoreCtx(DefaultRestoreFlags, &sb))
ctx.WriteKeyWord(sb.String())
```

Comment (kennytm) [reply ast/dml.go]: Both is fine as a workaround, and we need to add a TODO comment improving this no matter which we choose 🙃

Comment (leoppro) [reply ast/dml.go]: TODO what? make a unit expression and change the TiDB?

Comment (kennytm) [reply ast/dml.go]: @leoppro Yes.

**Comment by leoppro** [comment]
LGTM, I think we should merge this pr to advance progress. And it will be another issue about "Keyword expression". What do you think about it? @kennytm

**Comment by kennytm** [comment]
@leoppro The keyword expression issue has already been addressed in #142. This PR is not blocked by keyword expressions, I'd just like a TODO to make it not forgotten.

**Pull request merged**